cpu/stm32_common: add watchdog for stm32

fjmolinas opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->

This PR adds a simple watchdog implementation for stm32x boards. The api has been kept farely simple:

`* wdg_enable(void)`
`* wdg_disable(void)`
` * wdg_reset(void)`
 `* wdg_init(uint32_t time)`

For stm32 disable doesn't do anything since the watchdog can't be disabled in these platforms once enabled. It does freeze the wdg_timer on sleep modes for stm32l4. This function is added expecting that other platforms might have a disable function.

Pending support:

- [ ] window timer
- [ ] EWI (early wake up interrupt) (it's part of windowed timer support)

### Testing procedure

I have tested on stm32f0, stm32f4, stml321, stm32l1 & stm32l4. Haven't tested on stm32f2,3,7 but the workflow is exactly the same. There is a very simple application test that set-ups the timer with a modifiable reset_time.

`make BOARD=nucleo-l476rg -C tests/periph_wdg/ clean all flash -j term`

<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->
https://github.com/RIOT-OS/RIOT/pull/7374

### Issues/PRs references

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

Based on the wdt api discussed https://github.com/RIOT-OS/RIOT/pull/7374

aabadie commented on cpu/stm32_common/periph/wdg.c:
```diff
@@ -0,0 +1,168 @@
+
+#ifdef __cplusplus
+}
+#endif
```
missing newline

aabadie commented on cpu/stm32_common/periph/wdg.c:
```diff
@@ -0,0 +1,168 @@
+    return pre;
+}
+
+static uint16_t _find_reload_value(uint8_t pre, uint32_t rst_time){
```
The opening curly brace should on the next line

aabadie commented on cpu/stm32_common/periph/wdg.c:
```diff
@@ -0,0 +1,168 @@
+/**
+ * @brief   Macro for calculating WDT timings in us from enum values
+ */
+#define WDG_TIMING(pre, rel) \
```
Instead of a macro, you can use a `static inline` function.

aabadie commented on cpu/stm32_common/periph/wdg.c:
```diff
@@ -0,0 +1,168 @@
+ * @{
+ *
+ * @file        wdg.c
+ * @brief       Independent Watchdog timer for STM3L
```
s/STM3**L**/STM32/ ?

aabadie commented on cpu/stm32_common/periph/wdg.c:
```diff
@@ -0,0 +1,168 @@
+
+static uint8_t _find_prescaler(uint32_t rst_time){
+    /* Divide by the range to get power of 2 of the prescaler*/
+    uint16_t r = ( rst_time / IWDG_STEP_US );
```
The parenthesis are not needed.

aabadie commented on cpu/stm32_common/periph/wdg.c:
```diff
@@ -0,0 +1,168 @@
+    _iwdg_lock();
+}
+
+static uint8_t _find_prescaler(uint32_t rst_time){
```
I think this file require some uncrustify.

aabadie commented on tests/periph_wdg/README.md:
```diff
@@ -0,0 +1,7 @@
+Expected result
+===============
+If everything is working OK, device should reboot every 2 s,
```
"**the** device should reboot every 2s"

aabadie commented on tests/periph_wdg/main.c:
```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (C) 2019 INRIA
```
s/INRIA/Inria/

aabadie commented on tests/periph_wdg/main.c:
```diff
@@ -0,0 +1,44 @@
+    wdg_init(WDG_RESET_TIME);
+    wdg_enable();
+
+    while(1){};
```
space missing before `{}`

aabadie commented on tests/periph_wdg/main.c:
```diff
@@ -0,0 +1,44 @@
+#include "timex.h"
+
+#ifndef WDG_RESET_TIME
+#define WDG_RESET_TIME      (2U*US_PER_SEC)
```
I would keep 2U here, but move the multiplication with `US_PER_SEC` when calling `wdg_init`. Otherwise, you get this message when the board is booting:
`Application should reset every ~ 2000000 seconds` instead of `Application should reset every ~2 seconds`

aabadie commented on tests/periph_wdg/main.c:
```diff
@@ -0,0 +1,44 @@
+
+int main(void)
+{
+    printf("RIOT edg test application\n");
```
s/edg/wdg/

aabadie commented on drivers/include/periph/wdg.h:
```diff
@@ -0,0 +1,54 @@
+ */
+
+/**
+ * @defgroup    drivers_periph
```
The doxygen group should be `drivers_periph_wdg`

fjmolinas commented:
@aabadie comments addressed.

Regarding the python test. I had done one previously but is hard to test proper timing only threw a uart connection. To verify that it is working properly you must not only verify a reset occurred but also verify the timing. When working with ms the delay caused by prints and the time for the cpu to restart greatly affect the time you get first prints. Therefore capturing the output of and checking the time it was sent isn't viable. Using the in system time on the device isn't possible either since the device is resetting.

I think there are two options that could work:
- verify if on reset some property of the uart connection changes
- have some external pin whose state changes on reset which would allow to measure the reset time properly.

aabadie commented on cpu/stm32_common/periph/wdg.c:
```diff
@@ -0,0 +1,164 @@
+    /* Wait for register to be updated */
+    volatile int timeout = 10000;
+    while(IWDG->SR && timeout){
+        timeout--;
```
So when this times out, the watchdog is not correctly initialised IIUC. Why not simply returning with a 0 value (which is invalid) ?
This makes me think that the API could be changed:
```c
int wdg_init(uint32_t rst_time)
```
And the function returns 0 (use an enum) when the initialization is successful. It returns -1 (or an errno code) in case of a problem.
I think it's useless to return the rst_time set: since it's also the input value, having a successful return (0) means rst_time is correctly set (e.g the value returned by `_wdg_time(pre, rel);` is a good approximation of the input parameter).

What do you think ?

fjmolinas replied:
I can agree, it was useful when I wanted to verify that the approximation was "ok". Should I remove the function to get the actual programmed time then? Or leave it as ENABLE_DEBUG option?

fjmolinas commented:
@aabadie uncrustified files. If you look at the python test I make use of two fucntions to get the min and max value for the wdg timer, that way on the test I can test for out of range value and sepcify time limits according to the cpu. (for stm32 its allways the same but maybe nor for other platforms)?

Also I removed showing the actual tiem value for the wdg timer unless the DEBUG option is enabled. Should I just remove this?

What do you think?

kaspar030 commented:
```
wdg_init(uint32_t time)
```

If we change that to ```wdg_init(uint32_t min_time, uint32_t max_time)```, the API doesn't need to change when window support gets added to the implementation.

```
wdg_restart(void)
```

In the header that is called "wdg_reset()". That feels ambiguous. Does it reset the wdg device or does it reset the wdg counter? Maybe change to "wdg_kick()" or "wdg_trigger()"?

```
uint32_t wdg_min_timeout(void)
uint32_t wdg_max_timeout(void)
```

As the API as is doesn't support multiple wdg devices, can we get away with simple defines? The values should be known at compile time.

Other than that, two more questions:
- can we say something about the time base used for the watchdog? Is it always us, or is it ticks?
- what are the intended semantics for wdg_init() and wdg_enable()? Does init implicitly enable? I would suggest it shouldn't ...

fjmolinas commented:
@kaspar030 @aabadie 

> what are the intended semantics for wdg_init() and wdg_enable()? Does init implicitly enable? I would suggest it shouldn't ...

I would also say it doesn't, we could change it to wdg_setup() and wdg_enable(). That could make it more clear.

>If we change that to wdg_init(uint32_t min_time, uint32_t max_time), the API doesn't need to change when window support gets added to the implementation.

Agree, makes sense.

> can we say something about the time base used for the watchdog? Is it always us, or is it ticks?

By say something do you mean in the header? In the name of the function? Or discussing it here? I think it should be microsecond. Conversion from ticks to us should be done in the respective periph implementation. 

> If we change that to wdg_init(uint32_t min_time, uint32_t max_time), the API doesn't need to change when window support gets added to the implementation.

Agree, will do.

> As the API as is doesn't support multiple wdg devices, can we get away with simple defines? The values should be known at compile time

Yep had already talked off line to @aabadie about doing this.

> That feels ambiguous. Does it reset the wdg device or does it reset the wdg counter? Maybe change to "wdg_kick()" or "wdg_trigger()"?

I understand the confusion. How do you fill about this for the api:

`* wdg_enable(void) ------------------> wdg_start(void)`
`* wdg_disable(void) -----------------> wdg_stop(void)`
`* wdg_reset(void) -------------------> wdg_kick(void)`
`* wdg_init(uint32_t min_time) -------> wdg_setup(uint32_t min_time, uint32_t max_time)`

fjmolinas commented:
@aabadie addressed comments!

@kaspar030 ping!

aabadie commented on tests/periph_wdg/tests/01-run.py:
```diff
@@ -0,0 +1,61 @@
+from testrunner import run
+
+# We test only up to 10ms, with smaller times mcu doesn't have time to
+# print sysyem time before resetting
```
typo: should be sys**t**em instead of sys**y**em

fjmolinas replied:
fixed

aabadie commented on tests/periph_wdg/tests/01-run.py:
```diff
@@ -0,0 +1,61 @@
+        child.expect_exact("[wdg]: - ")
+        child.expect_exact(" -")
+
+        if child.before != "invalid configuration time":
```
You don't need this check. This (minor) case can simply be tested sending and invalid configuration using `sendline` and test the output is as expect using `expect_exact` or a regexp in `expect`.

fjmolinas replied:
Invalid configuration time will be different depending on each individual wdg.

I could move IWDG_MIN_TIME_US and IWDG_MAX_TIME_US to periph_cpu_common.h and therefore define invalid ranges from capturing so define values on start up. What do you think?

aabadie replied:
I think it's a good idea :) These values could be asserted as well when configuring the watchdog.

aabadie commented on tests/periph_wdg/tests/01-run.py:
```diff
@@ -0,0 +1,61 @@
+        if child.before != "invalid configuration time":
+            child.sendline("startloop")
+            start_time = get_start_time(child)
+            reset_time = get_reset_time(child)
```
I find `get_reset_time` overly complex. This could be much much simpler using `child.expect` and regex to get the reset time. I also don't like the use of json in the output of the application.
Have a look at how numerical values are retrieved using regexp in [the python script periph_rtt](https://github.com/RIOT-OS/RIOT/blob/master/tests/periph_rtt/tests/01-run.py#L20). This can be done the same way (not tested):
```python
child.expect(r'  Reset time: (\d+)')
reset_time = int(child.match.group(1))
```

Same for start time:
```python
child.expect(r'  Start time: (\d+)')
start_time = int(child.match.group(1))
```

fjmolinas replied:
@aabadie  Changed it accordingly, what do you think about it now?

aabadie commented on tests/periph_wdg/main.c:
```diff
@@ -0,0 +1,114 @@
+    { "kick", "Delay wdg timer", kick_wdg },
+    { "start", "Start wdg timer", start_wdg },
+    { "startloop", "Start wdg timer", start_loop_wdg },
+    { "stop", "Start wdg timer", stop_wdg },
```
Command description should be "Stop wdg timer"

aabadie commented on tests/periph_wdg/main.c:
```diff
@@ -0,0 +1,114 @@
+    { "setup", "Setup Wdg Timer", setup_wdg },
+    { "kick", "Delay wdg timer", kick_wdg },
+    { "start", "Start wdg timer", start_wdg },
+    { "startloop", "Start wdg timer", start_loop_wdg },
```
Command description is the same as the `start` command, copy-paste ?

fjmolinas commented on cpu/stm32_common/periph/wdg.c:
```diff
@@ -0,0 +1,156 @@
+{
+#ifdef CPU_FAM_STM32L4
+    DEBUG("[Warning]: wdg will freeze in STOP and STANDBY mode");
+    FLASH->OPTR |= ~(FLASH_OPTR_IWDG_STOP || FLASH_OPTR_IWDG_STDBY);
```
Note to self: this should be configured with a configurable define.

fjmolinas commented:
Rebased to make use of #11527.

benpicco commented on cpu/stm32_common/include/periph_cpu_common.h:
```diff
@@ -99,6 +99,23 @@ extern "C" {
+#if defined(CPU_FAM_STM32L4)
+#define HAS_WDT_INIT                   (1U)
+#else
+#define HAS_WDT_STOP                   (0U)
```
Did you mean `HAS_WDT_INIT `?

fjmolinas replied:
Yep, thanks for the catch. Addressed.

fjmolinas commented:
rebased and ping @aabadie

fjmolinas commented:
ping @aabadie

aabadie commented on cpu/stm32_common/periph/wdt.c:
```diff
@@ -0,0 +1,144 @@
+ * @author      Francisco Molina <francois-xavier.molina@inria.fr>
+ */
+
+#include "assert.h"
```
```suggestion
#include <assert.h>
```

aabadie commented on tests/periph_wdt/main.c:
```diff
@@ -0,0 +1,123 @@
+    return 0;
+}
+
+#if (HAS_WDT_STOP == 1)
```
```suggestion
#if HAS_WDT_STOP
```

aabadie commented on tests/periph_wdt/Makefile:
```diff
@@ -0,0 +1,20 @@
+
+USEMODULE += xtimer
+USEMODULE += shell
+USEMODULE += shell_commands
```
This is not needed I think.

aabadie commented:
Found some minor things but otherwise it looks good. I'll rebase #11556 on top of this to try the test application/script on nrf.

fjmolinas commented:
@aabadie Addressed, I also realized I had a typo in all the HAS_WDT_STOP, they weren't matching the header.

aabadie commented:
I like the status of this PR and I think it can be merged. Do you think you run the build and test script on your STM32 boards, but just for the wdt teat application ?
This way, this PR is tested on all STM32 families supported in RIOT.

fjmolinas commented:
Ran tests on one stm32 of each family we support. I had issues with some boards during flashing, there might be an issue with those boards. I'll re-check on Monday.

```
#### nucleo-f091rc/failuresummary.md

Failures during test.flash:
- [tests/periph_wdt](tests/periph_wdt/test.flash.failed)

#### nucleo-f103rb/failuresummary.md


#### nucleo-f207zg/failuresummary.md


#### nucleo-f303k8/failuresummary.md


#### nucleo-f446re/failuresummary.md


#### nucleo-f767zi/failuresummary.md

Failures during test:
- [tests/periph_wdt](tests/periph_wdt/test.failed)

#### nucleo-l073rz/failuresummary.md

Failures during test.flash:
- [tests/periph_wdt](tests/periph_wdt/test.flash.failed)

#### nucleo-l152re/failuresummary.md

#### nucleo-l432kc/failuresummary.md

```

fjmolinas commented:
@aabadie I have all tests passing except for `STM32F7`, related to #12181. But I have a bigger issue here, `IWDG` is using `CLOCK_LSI` which seems to have big deviations to its theoretical value, and it can't be calibrated, just measured. But this isn't possible unless connecting an oscilloscope to the board. So I think I need to relax the checked values, otherwise the test won't be deterministic since it will depend on the specific board `CLOCK_LSI`. What do you think? (e.g. for my nucleo-f767zi I seem to have a 15% deviation)

aabadie commented:
> What do you think?

Can you try again with #12181 merged ? (just on the F7)

aabadie commented:
> So I think I need to relax the checked values, otherwise the test won't be deterministic since it will depend on the specific board CLOCK_LSI. What do you think?

+1

fjmolinas commented:
I have been playing a little with these values and it depends highly on the board. With one f7 I have only 5% deviation and with another up to 30% deviation. Datasheet whise it seem deviation can go anywhere from "51% to 156%" of the value. So I changed margin of error to 50%. I also added a waring when compiling for `SMT32` so it is made explicit that LSI calibration is needed.

Tests is passing now on my `nucleo-f767zi` if a highly deviated LSI:

```
Entering directory '/home/francisco/workspace/RIOT/tests/periph_wdt'
PASSED target time: 100.0[ms], actual_time: 117.126[ms]
PASSED target time: 500.0[ms], actual_time: 579.362[ms]
PASSED target time: 1000.0[ms], actual_time: 1161.358[ms]
PASSED target time: 5000.0[ms], actual_time: 5817.32[ms]
TEST PASSED
```

Should

aabadie commented on cpu/stm32_common/periph/Makefile:
```diff
@@ -15,4 +15,9 @@ ifneq (,$(filter periph_flashpage periph_eeprom,$(USEMODULE)))
 endif
 
+ifneq (,$(filter periph_wdt,$(USEMODULE)))
+$(warning Attention! WDT is clocked by CLOCK_LSI, it needs manual measuring\
```
Please indent with 2 spaces inside the if block

aabadie commented on cpu/stm32_common/periph/Makefile:
```diff
@@ -15,4 +15,9 @@ ifneq (,$(filter periph_flashpage periph_eeprom,$(USEMODULE)))
 
+ifneq (,$(filter periph_wdt,$(USEMODULE)))
+$(warning Attention! WDT is clocked by CLOCK_LSI, it needs manual measuring\
+    since values can deviate upto 50% from reference)
```
put space between _up_ and _to_ ?

aabadie commented on tests/periph_wdt/README.md:
```diff
@@ -0,0 +1,11 @@
+## About
+
+This application allows testing the wdg peripheral. User can define the reset
+time threw the shell.
```
s/threw/through/ ?

aabadie commented on tests/periph_wdt/README.md:
```diff
@@ -0,0 +1,11 @@
+
+## Expected Result
+
+If wdg was initiated with a valid reset period then once it is started user
```
s/initiated/initialized/ ?

aabadie commented on tests/periph_wdt/Makefile:
```diff
@@ -0,0 +1,19 @@
+BOARD ?= nucleo-l152re
+include ../Makefile.tests_common
+
+TEST_ON_CI_WHITELIST += all
```
This line is not needed anymore

aabadie commented on tests/periph_wdt/main.c:
```diff
@@ -0,0 +1,123 @@
+{
+    (void)argc;
+    (void)argv;
+    puts("[wdt]: stoping wdt timer");
```
s/stoping/stopping/

aabadie commented on tests/periph_wdt/main.c:
```diff
@@ -0,0 +1,123 @@
+}
+
+static const shell_command_t shell_commands[] = {
+    { "range", "Return Wdg Timer range", get_range },
```
Here and line below Watchdog is called _Wdg_ but in other places you use _wdt_. Let's be consistent and use _wdt_ everywhere.

aabadie commented on tests/periph_wdt/main.c:
```diff
@@ -0,0 +1,123 @@
+    printf("[wdt]: start time: %" PRIu32 " us\n", xtimer_now_usec());
+    wdt_start();
+    while (1) {
+        printf("[wdt]: reset time: %" PRIu32 " us\n", xtimer_now_usec());
```
Prefixing with _[wdt]_ in the test application is not helpful to determine from the message comes from: the same prefix is used when enable debug in the peripheral implementation. Maybe just remove the prefix in the test application ? (and save some bytes on ROM).

aabadie commented on tests/periph_wdt/main.c:
```diff
@@ -0,0 +1,123 @@
+#include <stdlib.h>
+
+#include "periph/wdt.h"
+#include "periph_cpu.h"
```
I think this include can be removed, since it's might be already included by `periph/wdt.h`.

aabadie commented on tests/periph_wdt/Makefile:
```diff
@@ -0,0 +1,19 @@
+# Comment this out to disable code in RIOT that does safety checking
+# which is not needed in a production environment but helps in the
+# development process:
+DEVELHELP ?= 1
```
I think this is not needed, same for `QUIET`.

aabadie commented on tests/periph_wdt/tests/01-run.py:
```diff
@@ -0,0 +1,63 @@
+    reset_time = 0
+    try:
+        while True:
+            child.expect(u"\[wdt\]: reset time: (\\d+) us", timeout=1)
```
`(\d+)` works the same as `(\\d+)` (only `\`).

aabadie commented on tests/periph_wdt/tests/01-run.py:
```diff
@@ -0,0 +1,63 @@
+# We don't check for accuracy, only order of magnitude. Some MCU use an
+# an internal un-calibrated clock as reference which can deviate in
+# more than 50% from theoretical values (e.g STM32 board CLOCK_LSI)
+error_marge = 0.5
```
Maybe use `error_margin` for variable name ?

aabadie commented on tests/periph_wdt/tests/01-run.py:
```diff
@@ -0,0 +1,63 @@
+            wdt_reset_time = (reset_time_us - start_time_us) / 1e3
+
+            if wdt_reset_time < rst_time*(1 - 0.5) or \
+               wdt_reset_time > rst_time*(1 + 0.5) :
```
no need for space before `:`

aabadie commented on tests/periph_wdt/tests/01-run.py:
```diff
@@ -0,0 +1,72 @@
+        else:
+            # timeout
+            return reset_time
+            child.expect(u"reset time: (\d+) us", timeout=1)
```
Those 2 lines (under the return) are useless

fjmolinas commented:
@aabadie Thanks a lot for the review!
